### PR TITLE
[SYCLomatic] Enable the migration of kernel implicit synchronization by using in order queue event extension

### PR DESCRIPTION
**Pull request by intwanghao**

### CODE REVIEW

File: clang/lib/DPCT/AnalysisInfo.cpp [hunk 1]
```diff
@@ -6525,10 +6532,17 @@ KernelPrinter &KernelCallExpr::SubmitStmtsList::print(KernelPrinter &Printer) {
   printList(Printer, NdRangeList,
             "ranges to define ND iteration space for the kernel");
   printList(Printer, CommandGroupList, "helper variables defined");
+  if (ImplicitSyncFlag) {
+    Printer.line(
+        std::string(
+            "cgh.depends_on(dpct::get_current_device().get_last_events()") +
+        std::string(DefaultStreamFlag ? "" : "[0]") + std::string(");"));
```
Comment (ziranzha) [inline clang/lib/DPCT/AnalysisInfo.cpp]:
```suggestion
    Printer.line(
        "cgh.depends_on(dpct::get_current_device().get_last_events()"),
        DefaultStreamFlag ? "" : "[0]", ");");
```

**Review by ziranzha — APPROVED** [review]

File: clang/include/clang/DPCT/DPCTOptions.inc [hunk 2]
```diff
@@ -800,6 +800,14 @@ DPCT_ENUM_OPTION(
             "Experimental extension that allows map an address range onto "
             "multiple allocations of physical memory.",
             false),
+        DPCT_OPTION_ENUM_VALUE(
+            "in_order_queue_event",
```
Comment (zhimingwang36) [inline clang/include/clang/DPCT/DPCTOptions.inc]: in_order_queue_event ==>  in_order_queue_events

File: clang/test/dpct/kernel_implicit_sync.cu [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+// CHECK:          sycl::nd_range<3>(sycl::range<3>(1, 1, 1), sycl::range<3>(1, 1, 1)),
+// CHECK:          [=](sycl::nd_item<3> item_ct1) {
+// CHECK:            kernel(a);
+// CHECK:          });
+// CHECK:      });
+  kernel<<<1,1>>>(a);
+
+// CHECK:  s1->submit(
+// CHECK:      [&](sycl::handler &cgh) {
+// CHECK:        cgh.depends_on(dpct::get_current_device().get_last_events()[0]);
```
Comment (zhimingwang36) [inline clang/test/dpct/kernel_implicit_sync.cu]: cgh.depends_on(dpct::get_default_queue().ext_oneapi_get_last_event()

File: clang/runtime/dpct-rt/include/dpct/device.hpp [hunk 4]
```diff
@@ -582,6 +582,19 @@ class device_ext : public sycl::device {
     lock.lock();
   }
 
+  std::vector<sycl::event> get_last_events() {
```
Comment (zhimingwang36) [inline clang/runtime/dpct-rt/include/dpct/device.hpp]: Consider add a static function for current device,  just as get_default_queue();

**Review by zhimingwang36 — APPROVED** [review]

**Pull request merged**